src/parser/core/Analyzer.ts

```diff
@@ -50,3 +56,93 @@ class Analyzer extends EventSubscriber {
 }
 
 export default Analyzer;
+
+type ConstructedDependency<T> = T extends new (options: Options) => infer R ? R : never;
+
+type InjectedDependencies<Deps extends Dependencies> = {
+  [Key in keyof Deps]: ConstructedDependency<Deps[Key]>;
+};
+
+export enum FunctionType {
+  Statistic,
+  Suggestion,
+}
+
+type FunctionalEventFilter = EventFilter<any> | Array<EventFilter<any>>;
+
+function buildFunctionalAnalyzer<Deps extends Dependencies, Result = any>(
+  functionType: FunctionType,
+  metric: Metric<Result>,
+  eventFilter: FunctionalEventFilter = Events.any,
+  dependencies?: Deps,
+) {
+  const eventFilters: Array<EventFilter<any>> = Array.isArray(eventFilter)
+    ? eventFilter
+    : [eventFilter];
+
+  const deps = dependencies ?? {};
+
+  const analyzer = class extends Analyzer {
+    static dependencies = deps;
+
+    eventList: AnyEvent[] = [];
+
+    constructor(options: Options) {
+      super(options);
+
+      eventFilters.forEach((filter) => this.addEventListener(filter, this.appendEvent));
+    }
+
+    private appendEvent(event: AnyEvent) {
+      this.eventList.push(event);
```

Comment:
this does make a shallow copy of the event list, which is not ideal. I was lazy with the APL checker's `EventFilter` and left them at the default `EventFilter.any` even though they need something like `[EventFilter.cast, EventFilter.updatespellusable, EventFilter.begincast, ...]`. I need to find or write a function that applies an `EventFilter` to an array of events.